07_future_promise/homework/notification_sender.cxx

```diff
@@ -17,4 +31,69 @@ int main()
 	std::string message = "build failed";
 
 	// ...
+
+	std::vector<std::thread> threads;
+	std::vector<std::future<bool>> futures;
+
+	start = std::chrono::steady_clock::now();
+
+	// start threads
+	for ( std::string strDev : responsible_devs )
+	{
+		auto threadSendMail = []( std::string dev, std::string msg, std::promise<bool> promise )
+		{
+			try
+			{
+				// version 1.0
+				// bool bResult = libmail::send_mail(dev, msg);
+				// promise.set_value( bResult );
+
+				// version 2.0
+				libmail::send_mail(dev, msg);
+				promise.set_value( true );
+			}
+			catch( ... )
+			{
+				promise.set_exception( std::current_exception() );
+			}
+		};
+
+		std::promise<bool> p;
+		std::future<bool> f = p.get_future();
+
+		futures.push_back( std::move(f) );
+		threads.push_back( std::thread( threadSendMail, strDev, "message", std::move(p) ) );
+	}
+
+	// print results
+	for ( unsigned int i = 0; i < responsible_devs.size(); ++i )
+	{
+		try
+		{
+			// version 1.0
+			// bool bResult = futures[i].get();
+
+			// version 2.0
+			bool bResult = true;
+			futures[i].get();
+
+			std::cout << "> Sending mail to: " << responsible_devs[i] << " - ";
+			std::cout << ((bResult) ? "OK" : "FAIL");
+			std::cout << std::endl;
```

Comment:
tak naprawdę w tym przypadku to już będzie oznaczało, że zawsze jest poprawnie wysłany email, więc wypisywanie można uprościć.

Reply:
Zgoda dla biblioteki 2.0.
Zostawilem to w celu kompatybilnosci z 1.0/2.0.